Binaries/DafnyPrelude.bpl

```diff
@@ -939,6 +943,16 @@ axiom (forall<T> s: Seq T, i: int, v: T :: { Seq#Index(Seq#Build(s,v), i) }
 axiom (forall s: Seq Box, bx: Box, t: Ty :: { $Is(Seq#Build(s,bx),TSeq(t)) }
     $Is(s,TSeq(t)) && $IsBox(bx,t) ==> $Is(Seq#Build(s,bx),TSeq(t)));
 
+function Seq#Create(ty: Ty, heap: Heap, len: int, init: HandleType): Seq Box;
```

Comment:
Generally, the types at the Boogie level are much more coarse-grained than in Dafny. For example, the single Boogie type `Ref` includes references to objects of all Dafny types (and also includes the `null` value and references to unallocated objects).

`HandleType` is part of the translation of functions as first-class values in Dafny. This is necessary, because Boogie is first-order and does not allow functions as values. `HandleType` is more coarse-grained than the various arrow types in Dafny, and so it includes total, partial, and read-effectful functions as well as functions of all arities and parameter types. The Dafny translation into Boogie generates axioms about different arities on demand, so there will be more uses of `HandleType` in the Boogie program generated from a Dafny program that uses functions.

The Boogie type `Box` is used to collect all possible values in a Dafny program. Every type has an injection (and sometimes a bijection) into `Box`. This lets Dafny encode polymorphism in Boogie.

Like any compiler or translation into a more primitive language, the primitive language allows a lot of things that are not possible in the source language. For example, assembly code gives you arbitrary goto's. The translation is responsible for making appropriate use of the constructs and values in the target language.